src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp

```diff
@@ -70,49 +93,111 @@ double ShenandoahMmuTracker::process_time_seconds() {
   return 0.0;
 }
 
-ShenandoahMmuTracker::ShenandoahMmuTracker() :
-    _generational_reference_time_s(0.0),
-    _process_reference_time_s(0.0),
-    _collector_reference_time_s(0.0),
-    _mmu_periodic_task(new ShenandoahMmuTask(this)),
-    _mmu_average(10, ShenandoahAdaptiveDecayFactor) {
+void ShenandoahMmuTracker::help_record_concurrent(ShenandoahGeneration* generation, uint gcid, const char *msg) {
+  double current = os::elapsedTime();
+  _most_recent_gcid = gcid;
+  _most_recent_is_full = false;
+
+  if (gcid == 0) {
+    fetch_cpu_times(_most_recent_gc_time, _most_recent_mutator_time);
+
+    _most_recent_timestamp = current;
+  } else {
+    double gc_cycle_duration = current - _most_recent_timestamp;
+    _most_recent_timestamp = current;
+
+    double gc_thread_time, mutator_thread_time;
+    fetch_cpu_times(gc_thread_time, mutator_thread_time);
+    double gc_time = gc_thread_time - _most_recent_gc_time;
+    _most_recent_gc_time = gc_thread_time;
+    _most_recent_gcu = gc_time / (_active_processors * gc_cycle_duration);
+    double mutator_time = mutator_thread_time - _most_recent_mutator_time;
+    _most_recent_mutator_time = mutator_thread_time;
+    _most_recent_mu = mutator_time / (_active_processors * gc_cycle_duration);
+    log_info(gc, ergo)("At end of %s: GCU: %.1f%%, MU: %.1f%% for duration %.3fs",
+                       msg, _most_recent_gcu * 100, _most_recent_mu * 100, gc_cycle_duration);
+  }
 }
 
-ShenandoahMmuTracker::~ShenandoahMmuTracker() {
-  _mmu_periodic_task->disenroll();
-  delete _mmu_periodic_task;
+void ShenandoahMmuTracker::record_young(ShenandoahGeneration* generation, uint gcid) {
+  help_record_concurrent(generation, gcid, "Concurrent Young GC");
 }
 
-void ShenandoahMmuTracker::record(ShenandoahGeneration* generation) {
-  shenandoah_assert_control_or_vm_thread();
-  double collector_time_s = gc_thread_time_seconds();
-  double elapsed_gc_time_s = collector_time_s - _generational_reference_time_s;
-  generation->add_collection_time(elapsed_gc_time_s);
-  _generational_reference_time_s = collector_time_s;
+void ShenandoahMmuTracker::record_bootstrap(ShenandoahGeneration* generation, uint gcid, bool candidates_for_mixed) {
+  // Not likely that this will represent an "ideal" GCU, but doesn't hurt to try
+  help_record_concurrent(generation, gcid, "Bootstrap Old GC");
+  if (candidates_for_mixed) {
+    _doing_mixed_evacuations = true;
+  }
+  // Else, there are no candidates for mixed evacuations, so we are not going to do mixed evacuations.
+}
+
+void ShenandoahMmuTracker::record_old_marking_increment(ShenandoahGeneration* generation, uint gcid, bool old_marking_done,
+                                                        bool has_old_candidates) {
+  // No special processing for old marking
+  double duration = os::elapsedTime() - _most_recent_timestamp;
```

Comment:
We only seem to use `duration` here, none of the following lines produce any used values?

Reply:
my mistake.  i've corrected the report to make use of all relevant information (in expand-old-on-demand).  Thanks.